### PR TITLE
Introduce a Junit BeforeAllCallback

### PR DESCRIPTION
**Pull request by long-sean**
BeforeAllCallback should allow the use of the same running docker containers for all tests in suite

### CODE REVIEW

File: acceptance-tests/src/test/java/com/octopus/sdk/test/ServerResourceExtension.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+  protected static class ServerResource implements ExtensionContext.Store.CloseableResource {
+    private static final OctopusDeployServer server = OctopusDeployServerFactory.create();
+
+    public OctopusDeployServer getServer() {
+      return server;
+    }
+
+    @Override
+    public void close() throws Exception {
+      server.close();
```
Comment (rain-on) [inline acceptance-tests/src/test/java/com/octopus/sdk/test/ServerResourceExtension.java:61]: and remove from rootStore?

Comment (rain-on) [reply acceptance-tests/src/test/java/com/octopus/sdk/test/ServerResourceExtension.java:61]: and maybe set to null

File: acceptance-tests/src/test/java/com/octopus/sdk/test/BuildInformationAcceptanceTestOctopus.java [hunk 2]
```diff
@@ -42,7 +43,7 @@
 import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Test;
 
-public class BuildInformationAcceptanceTest extends BaseAcceptanceTest {
+public class BuildInformationAcceptanceTestOctopus extends BaseOctopusServerIntegrationTest {
```
Comment (rain-on) [inline acceptance-tests/src/test/java/com/octopus/sdk/test/BuildInformationAcceptanceTestOctopus.java:46]: Naming? Don't love that an "AcceptanceTest" inherits from an IntegrationTest.
Is BaseOctopusServerIntegrationTest maybe a "OctopusServerEnabledTest"? :shrug: might be overthinking it

File: test-support/src/main/java/com/octopus/testsupport/BaseOctopusServerIntegrationTest.java [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+  protected static OctopusDeployServer server;
+  protected static final OkHttpClient httpClient = new OkHttpClient();
+
+  @BeforeAll
+  public static void setup(final OctopusServerResourceExtension.ServerResource serverResource) {
+    server = serverResource.getServer();
+  }
+
+  protected void deleteSpaceValidly(
```
Comment (rain-on) [inline test-support/src/main/java/com/octopus/testsupport/BaseOctopusServerIntegrationTest.java:23]: not sure this lives here - i.e. the purpose of this class is to enforce the magic of a single OctpousServer - not offer random helpers.

File: test-support/src/main/java/com/octopus/testsupport/OctopusServerResourceExtension.java [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+import org.junit.jupiter.api.extension.ParameterResolutionException;
+import org.junit.jupiter.api.extension.ParameterResolver;
+
+public class OctopusServerResourceExtension implements BeforeAllCallback, ParameterResolver {
+
+  private static final Lock reentrantLock = new ReentrantLock();
+
+  @Override
+  public void beforeAll(final ExtensionContext context) {
+    reentrantLock.lock();
```
Comment (rain-on) [inline test-support/src/main/java/com/octopus/testsupport/OctopusServerResourceExtension.java:18]: is it worth making this a synchronize(this) block? looks cleaner than the lock (and removes the try-catch)

File: test-support/src/main/java/com/octopus/testsupport/OctopusServerResourceExtension.java [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+    try {
+      ExtensionContext.Store rootStore =
+          context.getRoot().getStore(ExtensionContext.Namespace.GLOBAL);
+      rootStore.put("octopus-server", new ServerResource());
```
Comment (rain-on) [inline test-support/src/main/java/com/octopus/testsupport/OctopusServerResourceExtension.java:22]: will this overwrite the existing entry for octopus-server? Is it meant to?

File: test-support/src/main/java/com/octopus/testsupport/BaseOctopusServerIntegrationTest.java (hunk 3 above)
Comment (long-sean) [reply test-support/src/main/java/com/octopus/testsupport/BaseOctopusServerIntegrationTest.java:23]: Agree to leave in place until test dsl is commenced

File: test-support/src/main/java/com/octopus/testsupport/BaseOctopusServerEnabledTest.java [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+
+import okhttp3.OkHttpClient;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.extension.ExtendWith;
+
+@ExtendWith(OctopusServerResourceExtension.class)
+public class BaseOctopusServerEnabledTest {
+
+  protected static OctopusDeployServer server;
+  protected static final OkHttpClient httpClient = new OkHttpClient();
```
Comment (rain-on) [inline test-support/src/main/java/com/octopus/testsupport/BaseOctopusServerEnabledTest.java:16]: not sure what this used to be, but doesn't need to be static

File: test-support/src/main/java/com/octopus/testsupport/OctopusServerResourceExtension.java [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+
+  private static final AtomicBoolean created = new AtomicBoolean(false);
+
+  @Override
+  public synchronized void beforeAll(final ExtensionContext context) {
+    if (created.compareAndSet(false, true)) {
+      ExtensionContext.Store rootStore =
+          context.getRoot().getStore(ExtensionContext.Namespace.GLOBAL);
+      final OctopusDeployServer server = OctopusDeployServerFactory.create();
+      rootStore.put("octopus-server", new ServerResource(server));
```
Comment (rain-on) [inline test-support/src/main/java/com/octopus/testsupport/OctopusServerResourceExtension.java:21]: "octopus-server" should probably be a constant which is reused in the ResolveParameter

**Review by rain-on — APPROVED** [review]
LGTM - 2 small ones

**Pull request merged**